### PR TITLE
Feature: C++11 API Modernisation

### LINKED ISSUE #3: Feature Request: Add support for range-based for loops to `JSONArray` and `JSONObject`
As it currently stands you are unable to use the range-based for loop with either the `JSONArray` object, or the `JSONObject` object. It would be a nice quality of life improvement if they both supported it.

For `JSONArray` the return of a `JSONAtom` is fine, and for `JSONObject` something like a `std::pair<decltype(key), JSONAtom>` would be nice.

### PR DESCRIPTION
**Pull request by dragonmux**
In this PR we overhaul the library's interface, making it a true C++11 library from the user's perspective, while going all-in on the internals being C++17.

This includes introducing proper iteration support on JSONObject, splitting and hiding the implementation details of JSONString, restructuring the code base, fixing CI, fixing several bugs including a couple of crashes caused by UB in the library, the implementation of a suite of new helpers on JSONAtom, JSONObject and JSONArray to make adding new elements to the two collection types simpler and less bug-prone, and fixing how the library exposes its object model to help the user apply correct ownership semantics to the objects returned and passed in.

This PR intentionally and deliberately breaks ABI with all previous versions of the library to affect these changes and fix the bugs that needed fixing.

This fixes #3

Items that still need addressing when this is done:

* ~~The sockets library needs its socket_t removing and replacing with substrate::socket_t~~ (done in this PR)
* The exceptions types need a major overhaul and making copy-safe
* JSONAtom's conversions need the C-style casts replacing with something safer
* The test suite needs rewriting onto crunch++
* The test suite needs expanding to cover all the new functionality properly

### CODE REVIEW

**Comment by dragonmux** [comment]
CC @amyspark as we'd appreciate her review on this

File: src/socket/meson.build [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	pkgconfig.generate(
+		rSONSocket,
+		name: 'rSON Socket',
+		filebase: 'rSON_socket',
+		libraries: [rSON],
+		version: meson.project_version(),
+		description: 'Sockets handling for rSON, a simple but effective JSON parser for C++'
+	)
+else
+	rSONSocket_dep = declare_dependency(
```
Comment (amyspark) [inline src/socket/meson.build:34]: I think this should be unconditional, that way you can consume all dependencies through `dependencies` and not `link_with` (which only brings the lib, and not the includes).

File: test/meson.build [hunk 2]
```diff
@@ -1,30 +1,61 @@
...
 
 if sanitizer != 'none'
-	commandExtra += ['-fsanitize=' + sanitizer]
+	command += ['-fsanitize=' + sanitizer]
+endif
+
+if not isWindows
+	testExt = '.so'
+else
+	testExt = '.dll'
```
Comment (amyspark) [inline test/meson.build:58]:
```suggestion
if isWindows
	testExt = '.dll'
else
	testExt = '.so'
```

File: test/meson.build [hunk 3]
```diff
@@ -50,22 +81,23 @@ testSrcs = [
 testHelpers = static_library(
 	'testHelpers',
 	testSrcs,
+	cpp_args: ['-DrSON_EXPORT_API'],
 	pic: true,
 	dependencies: [libcrunch, substrate],
-	include_directories: include_directories('..'),
+	include_directories: rSONIncludeDir,
```
Comment (amyspark) [inline test/meson.build:87]: Same as above regarding `dependencies` instead of `link_with`.

**Review by amyspark — COMMENTED** [review]
Hey all, I don't have much in the way of code comments (I've never used JSON) but I have a couple points that can be improved. Let me know what you think.

File: src/socket/meson.build (hunk 1 above)
Comment (dragonmux) [reply src/socket/meson.build:34]: Ah, good catch - in other words, line 33 becomes the `endif` and the rest de-indented, popping the dependency down out of the conditional block, correct?

Comment (dragonmux) [reply src/socket/meson.build]: We've implemented what we hope is an appropriate change for this - please give it a look over when you get a moment.

File: src/socket/meson.build [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	rSONSocketLibs += cxx.find_library('Ws2_32')
+endif
+
+rSONSocketSrc = ['socket.cxx']
+
+rSONSocket = library(
+	'rSON.socket',
+	rSONSocketSrc,
+	cpp_args: ['-DrSON_SOCKET_EXPORT_API'],
+	include_directories: [rSONIncludeDir, rSONSocketIncludeDir],
```
Comment (amyspark) [inline src/socket/meson.build:16]: @dragonmux you can drop `rSONIncludeDir` by virtue of the dep:

```suggestion
	include_directories: [rSONSocketIncludeDir],
```

**Review by amyspark — APPROVED** [review]

**Review by lethalbit — APPROVED** [review]
lgtm~ 🐱

File: src/socket/meson.build (hunk 4 above)
Comment (dragonmux) [reply src/socket/meson.build]: Good catch, thank you!

**Comment by dragonmux** [comment]
With Amy's approval, we'll pull the rSON::socket fixes from her onto this branch prior to merge once the substrate end of that has been resolved. We'll aim to merge this in the next 24h if possible - though no rush as we'd prefer to get the details right than doing a botched job.

**Pull request merged**